### PR TITLE
[dLLM] Cut per-step host overhead in the dLLM extend path

### PR DESCRIPTION
**Pull request by wenxuewuhd**
## Motivation

dLLM inference disables the overlap scheduler (each denoise step has a hard input dependency on the previous step's result), so every bit of host time between steps is exposed directly as device idle. Profiling LLaDA2 at bs=128 on Ascend 910B3 showed three host-side costs on the per-step path, none of which needed to be there:

1. `AscendAttnBackend.init_forward_metadata` took two D2H syncs to obtain values the scheduler already holds on the host (`seq_lens_max`, `extend_seq_lens_cpu_int`).
2. `NPUGraphRunner` re-read `forward_batch.seq_lens.cpu()` before every replay, a blocking D2H behind whatever is queued on the stream, while the same host mirror was already available.
3. `_alloc_extend_loc_with_kv_reuse` (the dLLM FDFO retained-KV path) re-read each reuse row's retained block with its own slice + cast kernel, and pulled every scheduling scalar out of CPU tensors one element at a time. Both costs grow linearly with batch size.

## Modifications

**`hardware_backend/npu/attention/ascend_backend.py`** — take `seq_lens_max` from `seq_lens_cpu` and build `extend_seq_lens_cpu_int` from `extend_seq_lens_cpu`. Gated on `forward_mode.is_dllm_extend()`, so the decode/prefill path is byte-identical.

**`hardware_backend/npu/graph_runner/npu_graph_runner.py`** — use `forward_batch.seq_lens_cpu` when present, falling back to `.cpu()` when it is not. This one is **not** dLLM-gated and affects every non-DeepSeek-DSA/V4 model on this path, so it is worth stating why it is equivalent: `seq_lens_cpu` is already the authoritative host mirror here — the Ascend attention backend derives its own `seq_lens_cpu_int` from it for every forward mode, and likewise adds the draft width on target-verify. Reading the device tensor in the graph runner made the two components disagree on their source for the same forward; this aligns them. The `is_target_verify()` adjustment is unchanged, and the two former branches collapse into one because they now differ only by that addition.

**`mem_cache/allocation.py`** — three changes inside `_alloc_extend_loc_with_kv_reuse`, which is reached only via `if batch.is_dllm()` in `alloc_for_extend`, so non-dLLM requests never enter it:

- Add `_gather_reused_block_locs`: the retained blocks all share one length, so fetch them with a single flat index op instead of one slice + cast kernel per row. It replaces one kernel launch per row with a fixed index build + H2D, so it wins on wide batches and loses on narrow ones. The gate is the reuse-row count, re-evaluated every round, so a batch that shrinks falls back on its own — a small-batch latency deployment keeps the per-row path without configuring anything. The threshold defaults to the backend's measured break-even (8 on Ascend, where break-even is ~6) and is unset on backends where it has not been measured, so those keep today's behavior. `SGLANG_DLLM_BATCHED_GATHER_MIN_ROWS` overrides it: a row count to retune, or `-1` to disable the batched path outright for A/B and rollback. Rows of non-uniform block length keep the per-row path regardless. The resolved default is logged once at INFO.
- Replace the per-element `int(cpu_tensor[i])` reads with one `.tolist()` per tensor. Unconditional: strictly less work at any batch size. At bs=128 the old code did roughly 770 of these, each allocating a temporary 0-dim tensor.
- Hoist the `torch.tensor([-1], device=...)` filler out of the `last_loc` comprehension. Unconditional for the same reason. Every empty-prefix row was allocating its own identical device tensor; the list is consumed by `torch.cat`, which copies, so one shared tensor is equivalent.

### Configuration

One new environment variable, `SGLANG_DLLM_BATCHED_GATHER_MIN_ROWS` (`EnvInt(None)` in `environ.py`), the minimum number of dLLM FDFO reuse rows at which the retained blocks are gathered in one flat index op. Nothing needs to be set for the intended behavior on either an Ascend or a non-Ascend deployment:

| Value | Behavior | When to use |
| --- | --- | --- |
| unset (default) | Ascend: batch at 8+ reuse rows. Other backends: always per-row. | Normal serving, both latency and throughput. |
| `N` (>= 0) | Batch at `N`+ reuse rows. | Enabling it on a backend after measuring its break-even, or retuning Ascend. |
| `-1` | Never batch. | Production A/B and quick rollback. |

```bash
# enable on a non-Ascend backend after measuring its break-even
SGLANG_DLLM_BATCHED_GATHER_MIN_ROWS=8 python3 -m sglang.launch_server ...

# roll back to the per-row path without redeploying
SGLANG_DLLM_BATCHED_GATHER_MIN_ROWS=-1 python3 -m sglang.launch_server ...
```

The resolved default is logged once at INFO:

```
dLLM retained-block gather: batching enabled above 8 reuse rows (Ascend default).
Override with SGLANG_DLLM_BATCHED_GATHER_MIN_ROWS.
```

## Accuracy Tests

No effect on model outputs — all three changes are host-side bookkeeping that produce the same values (the same slot indices, the same metadata) by different means. Verified with LLaDA2-mini on gsm8k (zero-shot, full 1319 samples): accuracy unchanged.

## Speed Tests and Profiling

Measured on Ascend 910B3 at bs=128 (LLaDA2-mini, 32-token dLLM blocks), microbenchmarking each site against the code it replaces:

| Site | Before | After | Saved |
| --- | ---: | ---: | ---: |
| Retained-block gather (128 reuse rows x 32) | 5671 us | 309 us | **5.4 ms** |
| ~770 scheduling scalar reads | 3849 us | 54 us | **3.3 ms** |
| `init_forward_metadata` (idle stream) | 251 us | 83 us | 168 us |
| Graph replay `seq_lens` (idle stream) | 44 us | 6 us | 38 us |

The gather is the dominant one and the only part that changes complexity: the per-row path costs about 44 us per reuse row (one kernel launch each) and scales linearly, while the batched path is roughly constant at about 270 us. That constant is why it is row-gated rather than unconditional — below the break-even the per-row path is genuinely faster:

| reuse rows | per-row | batched | |
| ---: | ---: | ---: | --- |
| 1 | 120 us | 305 us | per-row path taken |
| 4 | 196 us | 269 us | per-row path taken |
| 6 | 279 us | 271 us | break-even |
| 8 | 362 us | 268 us | Ascend default threshold |
| 32 | 1449 us | 277 us | |
| 64 | 2762 us | 268 us | |
| 128 | 5671 us | 309 us | |
| 256 | 11466 us | 353 us | |

The `last_loc` hoist is not in the table because its benefit depends on how many requests have an empty prefix: each such row was allocating a device tensor at about 70 us, so it saves that times the number of empty-prefix rows in the batch (up to 8.9 ms at bs=128 when every row is fresh, and nothing when every row has a cached prefix).

On Ascend at bs=128 this removes about 8.7 ms per step out of the box, against a step time of about 98 ms. On a backend where the gather threshold is unset it removes about 3.5 ms (the scalar reads plus the two D2H syncs); the gather itself is one env var away once its break-even has been measured there.

## Unit Tests

`test/registered/unit/mem_cache/test_dllm_reused_block_gather.py` (CPU-only, no server) pins the flat index arithmetic in `_gather_reused_block_locs` against per-row slicing. Each case covers a way the gather can be heterogeneous across rows, which is where a "looks equivalent" rewrite goes wrong: distinct per-row start offsets, unsorted pool indices, and a block ending exactly at the row boundary. Changing `row * width` to `row * block_len` in the implementation fails all three.

These are deliberately not a re-test of the happy path. `test_dllm_fdfo_kv_reuse.py` covers that already in principle, but its reuse rows all share one prefix length and ascending pool indices, so an implementation that broadcast row 0's offset across every row would still pass it. (It is also module-level skipped today, and its fixtures no longer match `alloc_for_extend` — un-skipping it raises `SimpleNamespace has no attribute out_cache_loc_dsv4` — so it is not currently guarding anything either way.)

The two D2H changes have no new tests: they need an NPU, and their equivalence is structural (`seq_lens_cpu` is the mirror of `seq_lens`, which the same backend already relies on).

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #35510470444](https://github.com/sgl-project/sglang/actions/runs/35510470444)
Latest PR Test (Extra): :x: [Run #35510470301](https://github.com/sgl-project/sglang/actions/runs/35510470301)
Latest PR Test (AMD ROCm 10): :x: [Run #35510470420](https://github.com/sgl-project/sglang/actions/runs/35510470420)

### CODE REVIEW

**Comment by rwang5203** [comment]
Black reformats the changed test, so [lint fails](https://github.com/sgl-project/sglang/actions/runs/30623561700/job/91133465554).

Run Black and commit the result.

File: test/registered/unit/mem_cache/test_dllm_reused_block_gather.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    def _gather(self, req_pool_indices, prefix_lens, block_len, dtype=torch.int64):
+        pool = _fingerprint_pool()
+        got = _gather_reused_block_locs(
+            req_to_token=pool,
+            req_pool_indices=req_pool_indices,
+            prefix_lens=prefix_lens,
+            block_len=block_len,
+            device=torch.device("cpu"),
+            out_dtype=dtype,
+        )
```
Comment (rwang5203) [inline test/registered/unit/mem_cache/test_dllm_reused_block_gather.py:66]: Helper-only tests miss batching and retained-row reassembly, allowing caller regressions.

Enable the integration test and cover mixed/all-reuse with batching on and off.

Comment (wenxuewuhd) [reply test/registered/unit/mem_cache/test_dllm_reused_block_gather.py:66]: thx for the suggestion.

Black: fixed.

Coverage: added cases against `_alloc_extend_loc_with_kv_reuse` itself. Each runs the same batch with the batched gather forced off and forced on and requires `out_cache_loc` to be identical — off is the pre-existing path, so agreement is the signal. Covers mixed reuse/fresh, all-reuse, unequal reuse block lengths, and a single reuse row with batching forced on. Advancing `reuse_ptr` on fresh rows, advancing `fresh_ptr` on reuse rows, or dropping the equal-`block_len` guard each turn a case red.

They vary `SGLANG_DLLM_BATCHED_GATHER_MIN_ROWS` rather than the auto default, since `_auto_min_rows_for_batched_gather` is `lru_cache`d.

If you meant enabling an existing test rather than adding these, say which one.

File: python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py [hunk 2]
```diff
@@ -443,7 +443,14 @@ def init_forward_metadata_out_graph(
     def init_forward_metadata(self, forward_batch: ForwardBatch):
         """Init the metadata for a forward pass."""
         self.forward_metadata = ForwardMetadata()
-        seq_lens_max = forward_batch.seq_lens.max()
+        # dLLM has no overlap schedule, so these two D2H syncs are exposed as
+        # device idle. Take them from the host mirrors instead. Gated on dLLM so
+        # the decode/prefill path stays byte-identical.
+        is_dllm = forward_batch.forward_mode.is_dllm_extend()
+        if is_dllm:
+            seq_lens_max = int(forward_batch.seq_lens_cpu.max())
```
Comment (rwang5203) [inline python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py:451]: The later fallback overwrites this host maximum with `seq_lens.max()`, retaining the synchronization this branch intends to remove.

Please preserve the host value for `DLLM_EXTEND` and test that this mode never reads the device reduction.

Comment (wenxuewuhd) [reply python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py]: The rebase pulled upstream's new trailing else: seq_lens_max = forward_batch.seq_lens.max() in under our hunk, turning the host read into a dead store. Fixed in 3360cc1: the dLLM case is now an elif is_dllm inside the chain, ahead of that fallback; target_verify and decode+spec are untouched. Added a unit test that runs init_forward_metadata for DLLM_EXTEND with seq_lens.max() rigged to raise, so the mode can never read the device reduction, plus a decode case pinning that path still reads it exactly once — both red without the fix.

The graph-runner change feeding seq_lens_cpu into replay_with_input_update stays ungated on purpose: the replay metadata path (init_forward_metadata_out_graph / build_replay_fb_view) already builds block tables from the same host mirror, so this keeps op attributes and block tables on one source instead of diverging under the overlap skew upstream's target-verify comment warns about. Current main has since moved in the same direction on its own — target_verify's seq_lens_max now comes from seq_lens_cpu upstream, and the verify offset already uses captured_req_width — so after rebasing, this PR's remaining graph-runner delta is just replacing the two leftover .cpu() D2H reads with that existing mirror.
